src/coprocessor/statistics/histogram.rs

```diff
@@ -0,0 +1,250 @@
+            ..Default::default()
+        }
+    }
+
+    pub fn into_proto(self) -> analyze::Histogram {
+        let mut hist = analyze::Histogram::new();
+        hist.set_ndv(self.ndv as i64);
+        let buckets: Vec<analyze::Bucket> = self.buckets
+            .into_iter()
+            .map(|bucket| bucket.into_proto())
+            .collect();
+        hist.set_buckets(RepeatedField::from_vec(buckets));
+        hist
+    }
+
+    // insert a data bigger than or equal to the max value in current histogram.
+    pub fn append(&mut self, data: Vec<u8>) {
+        if let Some(bucket) = self.buckets.last_mut() {
+            // The new item has the same value as last bucket value, to ensure that
+            // a same value only stored in a single bucket, we do not increase bucket
+            // even if it exceeds per_bucket_limit.
+            if bucket.upper_bound == data {
+                bucket.count_repeated();
+                return;
+            }
+        }
+        self.ndv += 1;
+        if self.buckets.len() >= self.buckets_num && self.is_last_bucket_full() {
+            self.merge_buckets();
+        }
+
+        if !self.is_last_bucket_full() {
+            self.buckets.last_mut().unwrap().append(data);
+            return;
+        }
+
+        // create a new bucket and insert data
+        let mut count = 1;
+        if let Some(bucket) = self.buckets.last() {
+            count += bucket.count
+        }
+        self.buckets.push(Bucket::new(count, data.clone(), data, 1));
+    }
+
+    // check whether the last bucket is full.
+    fn is_last_bucket_full(&self) -> bool {
+        if self.buckets.is_empty() {
+            return true;
+        }
+        let count = if self.buckets.len() == 1 {
+            self.buckets[0].count
+        } else {
+            let len = self.buckets.len();
+            self.buckets[len - 1].count - self.buckets[len - 2].count
+        };
+        count >= self.per_bucket_limit
+    }
+
+    // It merges every two neighbor buckets.
+    fn merge_buckets(&mut self) {
+        let bucket_num = (self.buckets_num + 1) / 2;
+        if bucket_num > 1 {
+            let (left, right) = self.buckets.split_at_mut(1);
+            mem::swap(&mut left[0].upper_bound, &mut right[0].upper_bound);
+            left[0].count = right[0].count;
+            left[0].repeats = right[0].repeats;
+        }
+        for id in 1..bucket_num {
+            let (left, right) = self.buckets.split_at_mut(id * 2);
+            if right.len() == 1 {
+                mem::swap(&mut left[id], &mut right[0]);
+                continue;
+            }
+            mem::swap(&mut left[id].lower_bound, &mut right[0].lower_bound);
+            mem::swap(&mut left[id].upper_bound, &mut right[1].upper_bound);
+            left[id].count = right[1].count;
+            left[id].repeats = right[1].repeats;
+        }
+        self.buckets.drain(bucket_num..);
+        self.per_bucket_limit *= 2;
```

Comment:
If `self.buckets_num == 1`, `merge` will causes `per_bucket_limit * 2` instead of **no effect**. Is it expected?

Reply:
yes @hicqu